src/ProxyClient/Http/HttpAsyncAdapter.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace FOS\HttpCache\ProxyClient\Http;
+
+use FOS\HttpCache\Exception\ProxyResponseException;
+use FOS\HttpCache\ProxyClient\Http\InvalidationRequest;
+use FOS\HttpCache\ProxyClient\Http\RequestQueue;
+use Http\Client\HttpAsyncClient;
+use Http\Client\HttpClient;
+use Http\Client\Promise;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * An adapter to work with the Httplug asynchronous client.
+ *
+ * @author David Buchmann <mail@davidbu.ch>
+ */
+class HttpAsyncAdapter extends HttpAdapterBase
+{
+    /**
+     * @var HttpAsyncClient
+     */
+    private $httpClient;
+
+    /**
+     * Asynchronously sent requests
+     *
+     * @var Promise[]
+     */
+    private $promises;
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param HttpAsyncClient $httpClient
+     */
+    public function __construct(array $servers, $baseUri, HttpAsyncClient $httpClient)
+    {
+        parent::__construct($servers, $baseUri);
+
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Immediately send the request asynchronously
+     */
+    public function invalidate(InvalidationRequest $invalidationRequest)
+    {
+        $signature = $invalidationRequest->getSignature();
+
+        if (isset($this->promises[$signature])) {
+            return;
+        }
+
+        $this->promises[$signature] = $this->httpClient->sendAsyncRequest($invalidationRequest);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Wait for all promises to resolve and gather exceptions.
+     */
+    public function flush()
+    {
+        $promises = $this->promises;
+        $this->promises = [];
+        foreach ($promises as $promise) {
+            $promise->wait();
+            // TODO: gather exceptions
+            // Handle all networking errors: php-http only throws an exception
```

Comment:
actually we can use a plugin in the chain to make http error responses exceptions. we probably should ask the user to do that and this will become easy.